File: scripts/download_crowdin.py
#!/usr/bin/env python3
"""Download Pekzep translations from Crowdin via the API (no CLI path mapping)."""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
OUT_DIR = ROOT / "translations"
OUT_FILE = OUT_DIR / "pz_ai.json"
PROJECT_ID = os.environ.get("CROWDIN_PROJECT_ID", "923393")
BASE = "https://api.crowdin.com/api/v2"


def token() -> str:
    value = os.environ.get("CROWDIN_PERSONAL_TOKEN", "").strip()
    if not value:
        raise SystemExit("CROWDIN_PERSONAL_TOKEN is not set")
    return value


def request(method: str, path: str, body: dict | None = None) -> dict:
    url = BASE + path
    data = None
    headers = {
        "Authorization": f"Bearer {token()}",
        "Accept": "application/json",
    }
    if body is not None:
        data = json.dumps(body).encode("utf-8")
        headers["Content-Type"] = "application/json"
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=60) as resp:
            raw = resp.read()
    except urllib.error.HTTPError as e:
        detail = e.read().decode("utf-8", errors="replace")
        raise SystemExit(f"Crowdin API {e.code} {method} {path}: {detail}") from e
    if not raw:
        return {}
    return json.loads(raw.decode("utf-8"))


def fetch_url(url: str) -> bytes:
    req = urllib.request.Request(url, headers={"Authorization": f"Bearer {token()}"})
    with urllib.request.urlopen(req, timeout=60) as resp:
        return resp.read()


def paginate(path: str) -> list:
    items: list = []
    offset = 0
    while True:
        sep = "&" if "?" in path else "?"
        payload = request("GET", f"{path}{sep}limit=500&offset={offset}")
        chunk = payload.get("data", [])
        items.extend(chunk)
        pagination = payload.get("pagination") or {}
        offset += pagination.get("limit", len(chunk))
        if offset >= pagination.get("total", offset) or not chunk:
            break
    return items


def unwrap(item: dict) -> dict:
    return item.get("data", item)


def pick_language(project: dict) -> str:
    source = unwrap(project).get("sourceLanguageId") or "en"
    targets = unwrap(project).get("targetLanguageIds") or []
    print(f"source language: {source}")
    print(f"target languages: {targets}")
    if not targets:
        raise SystemExit("Crowdin project has no target languages")
    pz = [lang for lang in targets if "pz" in lang.lower()]
    if len(pz) == 1:
        return pz[0]
    remaining = [lang for lang in targets if lang != source]
    if len(remaining) == 1:
        return remaining[0]
    if len(pz) > 1:
        raise SystemExit(f"multiple pz languages: {pz}")
    raise SystemExit(f"cannot choose target language from {targets}")


def main() -> int:
    project = request("GET", f"/projects/{PROJECT_ID}")
    language = pick_language(project)
    print(f"downloading language: {language}")

    files = [unwrap(item) for item in paginate(f"/projects/{PROJECT_ID}/files")]
    if not files:
        raise SystemExit("Crowdin project has no files")
    for info in files:
        print(f"file id={info.get('id')} path={info.get('path')} name={info.get('name')}")

    merged: dict = {}
    for info in files:
        file_id = info["id"]
        built = request(
            "POST",
            f"/projects/{PROJECT_ID}/translations/builds/files/{file_id}",
            {
                "targetLanguageId": language,
                "skipUntranslatedStrings": False,
                "skipUntranslatedFiles": False,
                "exportApprovedOnly": False,
            },
        )
        url = unwrap(built).get("url")
        if not url:
            raise SystemExit(f"no download url for file {file_id}: {built}")
        raw = fetch_url(url)
        try:
            payload = json.loads(raw.decode("utf-8"))
        except json.JSONDecodeError as e:
            raise SystemExit(
                f"translation for {info.get('path')} is not JSON: {raw[:200]!r}"
            ) from e
        if not isinstance(payload, dict):
            raise SystemExit(f"translation for {info.get('path')} is not a JSON object")
        merged.update(payload)
        print(f"merged {len(payload)} keys from {info.get('path')}")

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    with OUT_FILE.open("w", encoding="utf-8", newline="\n") as f:
        json.dump(merged, f, ensure_ascii=False, indent=2)
        f.write("\n")
    print(f"wrote {OUT_FILE.relative_to(ROOT)} ({len(merged)} keys)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/download_crowdin.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");

const ROOT = path.resolve(__dirname, "..");
const OUT_DIR = path.join(ROOT, "translations");
const OUT_FILE = path.join(OUT_DIR, "pz_ai.json");
const PROJECT_ID = process.env.CROWDIN_PROJECT_ID || "923393";
const BASE = "https://api.crowdin.com/api/v2";
const TIMEOUT_MS = 60_000;

function token() {
  const value = (process.env.CROWDIN_PERSONAL_TOKEN || "").trim();
  if (!value) {
    throw new Error("CROWDIN_PERSONAL_TOKEN is not set");
  }
  return value;
}

async function request(method, apiPath, body) {
  const headers = {
    Authorization: `Bearer ${token()}`,
    Accept: "application/json",
  };
  const opts = { method, headers, signal: AbortSignal.timeout(TIMEOUT_MS) };
  if (body !== undefined) {
    headers["Content-Type"] = "application/json";
    opts.body = JSON.stringify(body);
  }
  const resp = await fetch(BASE + apiPath, opts);
  const raw = Buffer.from(await resp.arrayBuffer());
  if (!resp.ok) {
    throw new Error(
      `Crowdin API ${resp.status} ${method} ${apiPath}: ${raw.toString("utf8")}`
    );
  }
  if (raw.length === 0) {
    return {};
  }
  return JSON.parse(raw.toString("utf8"));
}

async function fetchUrl(url) {
  // Crowdin artifact URLs are signed; sending the API token makes S3 return 400.
  const resp = await fetch(url, { signal: AbortSignal.timeout(TIMEOUT_MS) });
  const raw = Buffer.from(await resp.arrayBuffer());
  if (!resp.ok) {
    throw new Error(
      `download ${resp.status} ${url.split("?", 1)[0]}: ${raw.toString("utf8")}`
    );
  }
  return raw;
}

async function paginate(apiPath) {
  const items = [];
  let offset = 0;
  for (;;) {
    const sep = apiPath.includes("?") ? "&" : "?";
    const payload = await request(
      "GET",
      `${apiPath}${sep}limit=500&offset=${offset}`
    );
    const chunk = payload.data || [];
    items.push(...chunk);
    const pagination = payload.pagination || {};
    offset += pagination.limit ?? chunk.length;
    if (offset >= (pagination.total ?? offset) || chunk.length === 0) {
      break;
    }
  }
  return items;
}

function unwrap(item) {
  return item && item.data !== undefined ? item.data : item;
}

function pickLanguage(project) {
  const source = unwrap(project).sourceLanguageId || "en";
  const targets = unwrap(project).targetLanguageIds || [];
  console.log(`source language: ${source}`);
  console.log(`target languages: ${JSON.stringify(targets)}`);
  if (targets.length === 0) {
    throw new Error("Crowdin project has no target languages");
  }
  const pz = targets.filter((lang) => lang.toLowerCase().includes("pz"));
  if (pz.length === 1) {
    return pz[0];
  }
  const remaining = targets.filter((lang) => lang !== source);
  if (remaining.length === 1) {
    return remaining[0];
  }
  if (pz.length > 1) {
    throw new Error(`multiple pz languages: ${JSON.stringify(pz)}`);
  }
  throw new Error(`cannot choose target language from ${JSON.stringify(targets)}`);
}

async function main() {
  const project = await request("GET", `/projects/${PROJECT_ID}`);
  const language = pickLanguage(project);
  console.log(`downloading language: ${language}`);

  const files = (await paginate(`/projects/${PROJECT_ID}/files`)).map(unwrap);
  if (files.length === 0) {
    throw new Error("Crowdin project has no files");
  }
  for (const info of files) {
    console.log(`file id=${info.id} path=${info.path} name=${info.name}`);
  }

  const merged = {};
  for (const info of files) {
    const fileId = info.id;
    const built = await request(
      "POST",
      `/projects/${PROJECT_ID}/translations/builds/files/${fileId}`,
      {
        targetLanguageId: language,
        skipUntranslatedStrings: false,
        skipUntranslatedFiles: false,
        exportApprovedOnly: false,
      }
    );
    const url = unwrap(built).url;
    if (!url) {
      throw new Error(`no download url for file ${fileId}: ${JSON.stringify(built)}`);
    }
    const raw = await fetchUrl(url);
    let payload;
    try {
      payload = JSON.parse(raw.toString("utf8"));
    } catch (err) {
      throw new Error(
        `translation for ${info.path} is not JSON: ${raw.subarray(0, 200)}`
      );
    }
    if (payload === null || typeof payload !== "object" || Array.isArray(payload)) {
      throw new Error(`translation for ${info.path} is not a JSON object`);
    }
    Object.assign(merged, payload);
    console.log(`merged ${Object.keys(payload).length} keys from ${info.path}`);
  }

  fs.mkdirSync(OUT_DIR, { recursive: true });
  fs.writeFileSync(OUT_FILE, JSON.stringify(merged, null, 2) + "\n", "utf8");
  console.log(
    `wrote ${path.relative(ROOT, OUT_FILE)} (${Object.keys(merged).length} keys)`
  );
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});
